lib/api/api_auth_webauthn.go

```diff
@@ -0,0 +1,472 @@
+		if err != nil {
+			badRequest, ok := err.(*webauthnProtocol.Error)
+			if ok && badRequest.Type == "invalid_request" && badRequest.Details == "Found no credentials for user" {
+				sendJSON(w, make(map[string]string))
+			} else {
+				l.Warnln("Failed to initialize WebAuthn login", err)
+			}
+			return
+		}
+
+		var req startWebauthnAuthenticationResponse
+		req.Options = *options
+		req.RequestID = uuid.New().String()
+		s.authenticationStates[req.RequestID] = s.startTimedSessionData(sessionData)
+
+		sendJSON(w, req)
+	}
+}
+
+func (s *webauthnService) finishWebauthnAuthentication(tokenCookieManager *tokenCookieManager, guiCfg config.GUIConfiguration) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer s.deleteOldStates()
+
+		var req finishWebauthnAuthenticationRequest
+
+		if err := unmarshalTo(r.Body, &req); err != nil {
+			l.Debugln("Failed to parse response:", err)
+			http.Error(w, "Failed to parse response.", http.StatusBadRequest)
+			return
+		}
+
+		state, ok := s.authenticationStates[req.RequestID]
+		if !ok {
+			l.Debugf("Unknown request ID: %s", req.RequestID)
+			badRequest(w)
+			return
+		}
+		delete(s.authenticationStates, req.RequestID) // Allow only one attempt per challenge
+
+		if s.expired(&state) {
+			l.Debugln("WebAuthn authentication timed out: %v", state)
+			http.Error(w, "Request Timeout", http.StatusRequestTimeout)
+			return
+		}
+
+		parsedResponse, err := req.Credential.Parse()
+		if err != nil {
+			l.Debugln("Failed to parse WebAuthn authentication response", err)
+			badRequest(w)
+			return
+		}
+
+		updatedCred, err := s.engine.ValidateLogin(s.user(guiCfg), state.sessionData, parsedResponse)
+		if err != nil {
+			l.Infoln("WebAuthn authentication failed", err)
+
+			if state.sessionData.UserVerification == webauthnProtocol.VerificationRequired {
+				antiBruteForceSleep()
+				http.Error(w, "Conflict", http.StatusConflict)
+				return
+			}
+
+			forbidden(w)
+			return
+		}
+
+		authenticatedCredId := base64.RawURLEncoding.EncodeToString(updatedCred.ID)
+		persistentState := guiCfg.WebauthnState
+
+		for _, cred := range persistentState.Credentials {
+			if cred.ID == authenticatedCredId {
+				if cred.RequireUv && !updatedCred.Flags.UserVerified {
+					antiBruteForceSleep()
+					http.Error(w, "Conflict", http.StatusConflict)
+					return
+				}
+				break
+			}
+		}
+
+		var signCountBefore uint32 = 0
+
+		volState := s.loadVolatileState()
+		dynCredState, ok := volState.Credentials[authenticatedCredId]
+		if !ok {
+			dynCredState = WebauthnCredentialVolatileState{}
+		}
+		signCountBefore = dynCredState.SignCount
+		dynCredState.SignCount = updatedCred.Authenticator.SignCount
+		dynCredState.LastUseTime = s.timeNow().Truncate(time.Second).UTC()
+		volState.Credentials[authenticatedCredId] = dynCredState
+		err = s.storeVolatileState(volState)
+		if err != nil {
+			l.Warnln("Failed to update authenticated WebAuthn credential", err)
+		}
```

Comment:
This could be factored out into a method on `*webauthnService`:

    func (s *webauthnService) updateVolatileSignCount(id string) error

And necessary locking could be handled in there.